File: src/main/java/io/github/binout/soccer/domain/season/SeasonStatistics.java
package io.github.binout.soccer.domain.season;

import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.domain.season.match.FriendlyMatch;
import io.github.binout.soccer.domain.season.match.LeagueMatch;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class SeasonStatistics {

    private final Map<String, Long> friendlyMatchesPlayed;
    private final Map<String, Long> leagueMatchesPlayed;
    private final long nbPlayers;

    SeasonStatistics(Season season) {
        List<String> friendlyPlayerGames = season.friendlyMatches().flatMap(FriendlyMatch::players).collect(Collectors.toList());
        this.friendlyMatchesPlayed = playersByGamePlayed(friendlyPlayerGames);
        List<String> leaguePlayerGames = season.leagueMatches().flatMap(LeagueMatch::players).collect(Collectors.toList());
        this.leagueMatchesPlayed = playersByGamePlayed(leaguePlayerGames);
        this.nbPlayers = Stream.concat(friendlyPlayerGames.stream(), leaguePlayerGames.stream()).distinct().count();

    }

    private Map<String, Long> playersByGamePlayed(List<String> allPlayerGames) {
        return allPlayerGames.stream().distinct().collect(Collectors.toMap(
                Function.identity(),
                p -> allPlayerGames.stream().filter(p::equals).count()));
    }

    public long nbPlayers() {
        return this.nbPlayers;
    }

    public int matchPlayed(Player player) {
        return leagueMatchPlayed(player) + friendlyMatchPlayed(player);
    }

    public int leagueMatchPlayed(Player player) {
        return this.leagueMatchesPlayed.getOrDefault(player.name(), 0L).intValue();
    }

    public int friendlyMatchPlayed(Player player) {
        return this.friendlyMatchesPlayed.getOrDefault(player.name(), 0L).intValue();
    }
}

File: src/main/kotlin/io/github/binout/soccer/domain/season/SeasonStatistics.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.domain.season

import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.player.PlayerName

class SeasonStatistics(season: Season) {

    private val friendlyMatchesPlayed: Map<PlayerName, Int>
    private val leagueMatchesPlayed: Map<PlayerName, Int>
    val nbPlayers: Int

    init {
        val friendlyPlayerGames = season.friendlyMatches().flatMap { it.players() }
        this.friendlyMatchesPlayed = playersByGamePlayed(friendlyPlayerGames)
        val leaguePlayerGames = season.leagueMatches().flatMap { it.players() }
        this.leagueMatchesPlayed = playersByGamePlayed(leaguePlayerGames)
        this.nbPlayers = (friendlyPlayerGames + leaguePlayerGames).distinct().count()
    }

    private fun playersByGamePlayed(allPlayerGames: List<PlayerName>): Map<PlayerName, Int> =
            allPlayerGames.distinct().associate { player -> player to allPlayerGames.count { player == it  } }

    fun matchPlayed(player: Player): Int = leagueMatchPlayed(player) + friendlyMatchPlayed(player)

    fun leagueMatchPlayed(player: Player): Int =
            this.leagueMatchesPlayed.getOrDefault(player.name, 0)

    fun friendlyMatchPlayed(player: Player): Int =
            this.friendlyMatchesPlayed.getOrDefault(player.name, 0)
}
